dspy/utils/caching.py

```diff
@@ -12,3 +18,58 @@ def create_subdir_in_cachedir(subdir: str) -> str:
     subdir = os.path.abspath(subdir)
     os.makedirs(subdir, exist_ok=True)
     return subdir
+
+
+class LRUCache(OrderedDict):
+    maxsize: int
+
+    def __init__(self, iterable, maxsize: int):
+        super().__init__(iterable)
+        self.maxsize = maxsize
+
+    @classmethod
+    def load(cls, file, maxsize: int) -> "LRUCache":
+        return cls(pickle.load(file), maxsize)
+
+    @staticmethod
+    def dump(obj, file) -> None:
+        pickle.dump([[k, v] for k, v in obj.items()], file)
+
+    def __setitem__(self, request: dict, value):
+        key = self.cache_key(request)
+
+        if key in self:
+            self.move_to_end(key)
+            return
+
+        if len(self) == self.maxsize:
+            self.popitem(last=False)
+
+        super().__setitem__(key, value)
+
+    def __getitem__(self, request: dict):
+        key = self.cache_key(request)
+        return super().__getitem__(key)
+
+    def __contains__(self, request: dict):
+        key = self.cache_key(request)
+        return super().__contains__(key)
+
+    def get(self, request: dict, default=None):
+        key = self.cache_key(request)
+        return super().get(key, default)
+
+    def __delitem__(self, request: dict):
+        key = self.cache_key(request)
+        super().__delitem__(key)
+
+    def pop(self, request: dict, default=None):
+        key = self.cache_key(request)
+        return super().pop(key, default)
+
+    @staticmethod
+    def cache_key(request: Union[dict, str]) -> str:
+        params = request
+        if isinstance(request, dict):
+            params = {k: v for k, v in request.items() if not callable(v)}
+        return sha256(ujson.dumps(params, sort_keys=True).encode()).hexdigest()
```

Comment:
@CyrusNuevoDia @bahtman Can we handle pydantic models here as well, by converting them to JSON objects? This will be necessary for supporting structured outputs - https://github.com/stanfordnlp/dspy/pull/1881/files#diff-ee21f597435b8e775fd162c84aaa2c7a544120407baa26c12b3ff26d0d31a7dbR43-R44

Reply:
Would the pydantic model be a part of the dictionary request object or the entirety of it? Do we want to cache it or just avoid it like the callable parameters? (Above implementation ignores pydantic models because of the callable check)

Reply:
Great question! I think we should cache it, I.e. it should be included in `params`. It will be part of the dictionary request object, e.g. an entry like `params["response_format"] = <Pydantic object>`

Reply:
Can we assume its always "response_format" and we just check for that and convert it to dict?

Reply:
I think it'd be a bit more future proof (and also totally sufficient) to check for `isinstance(v, pydantic.BaseModel)`

Reply:
To avoid another passthrough I would prefer a staticmethod `convert_pydantic` that check for `isinstance(v, pydantic.BaseModel)` and if true returns `v.schema()` else `v`. Then we can just call that while we do the dict comprehension.

Reply:
`params = {k: self.convert_pydantic(v) for k, v in request.items() if not callable(v)}`

Reply:
Ahh. This might give us issues, since v is callable when a pydantic. Either we build an or statement or we just do 2 passthroughs (Later is probably more clean at this point)